modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/util/ProjectedTupleTest.java

```diff
@@ -0,0 +1,175 @@
+    void testProjection(boolean useOptimizeProjection) {
+        int f1 = RND.nextInt(ALL_TYPES_SCHEMA.elementCount());
+        int f2 = RND.nextInt(ALL_TYPES_SCHEMA.elementCount());
+        int f3 = RND.nextInt(ALL_TYPES_SCHEMA.elementCount());
+
+        int[] projection = {f1, f2, f3};
+
+        InternalTuple projectedTuple = useOptimizeProjection
+                ? new ProjectedTuple(TUPLE, projection)
+                : new ProjectedTuple(ALL_TYPES_SCHEMA, TUPLE, projection);
+
+        Element e1 = ALL_TYPES_SCHEMA.element(f1);
+        Element e2 = ALL_TYPES_SCHEMA.element(f2);
+        Element e3 = ALL_TYPES_SCHEMA.element(f3);
+
+        BinaryTupleSchema projectedSchema = BinaryTupleSchema.create(new Element[] {
+                e1, e2, e3
+        });
+
+        assertThat(projectedSchema.value(projectedTuple, 0), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f1)));
+        assertThat(projectedSchema.value(projectedTuple, 1), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f2)));
+        assertThat(projectedSchema.value(projectedTuple, 2), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f3)));
+
+        InternalTuple restored = new BinaryTuple(projection.length, projectedTuple.byteBuffer());
+
+        assertThat(projectedSchema.value(restored, 0), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f1)));
+        assertThat(projectedSchema.value(restored, 1), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f2)));
+        assertThat(projectedSchema.value(restored, 2), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f3)));
+    }
+
+    private static NativeType fromElement(Element element) {
+        switch (element.typeSpec()) {
+            case BOOLEAN:
+                return NativeTypes.BOOLEAN;
+            case INT8:
+                return NativeTypes.INT8;
+            case INT16:
+                return NativeTypes.INT16;
+            case INT32:
+                return NativeTypes.INT32;
+            case INT64:
+                return NativeTypes.INT64;
+            case FLOAT:
+                return NativeTypes.FLOAT;
+            case DOUBLE:
+                return NativeTypes.DOUBLE;
+            case DECIMAL:
+                return NativeTypes.decimalOf(20, element.decimalScale());
+            case NUMBER:
+                return NativeTypes.numberOf(20);
+            case DATE:
+                return NativeTypes.DATE;
+            case TIME:
+                return NativeTypes.time();
+            case DATETIME:
+                return NativeTypes.datetime();
+            case TIMESTAMP:
+                return NativeTypes.timestamp();
+            case UUID:
+                return NativeTypes.UUID;
+            case BITMASK:
+                return NativeTypes.bitmaskOf(256);
+            case STRING:
+                return NativeTypes.stringOf(256);
+            case BYTES:
+                return NativeTypes.blobOf(256);
+            default:
+                throw new IllegalArgumentException("Unknown type: " + element.typeSpec());
+        }
+    }
+}
```

Comment:
We usually put an empty line at the end of the file.

Reply:
good catch! fixed